QUESTION NUMBER 9, PAGE 227-228, CHAPTER 5 OF THE TEXTBOOK "AN INTRODUCTION TO STATISTICAL LEARNING WITH APPLICATIONS IN PYTHON".

We will now consider the Boston housing data set, from the ISLP library.

A data set containing housing values in 506 suburbs of Boston.
Column   Description
crim     per capita crime rate by town.
zn       proportion of residential land zoned for lots over 25,000 sq.ft.
indus    proportion of non-retail business acres per town.
chas     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise).
nox      nitrogen oxides concentration (parts per 10 million).
rm       average number of rooms per dwelling.
age      proportion of owner-occupied units built prior to 1940.
dis      weighted mean of distances to five Boston employment centres.
rad      index of accessibility to radial highways.
tax      full-value property-tax rate per $10,000.
ptratio  pupil-teacher ratio by town.
lstat    lower status of the population (percent).
medv     median value of owner-occupied homes in $1000s.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from ISLP import load_data

boston = load_data("Boston")

QUESTION A

Based on this data set, provide an estimate for the population mean of medv. Call this estimate µˆ.

In [2]:
# Assume df is loaded and has column 'medv'
medv = boston['medv'].to_numpy()
n = len(medv)

mu_hat = np.mean(medv)

print(f"Mean (mu_hat): {mu_hat:.3f}")


Mean (mu_hat): 22.533


QUESTION B

Provide an estimate of the standard error of µˆ. Interpret this result. 
Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [3]:
se_mean_analytic = np.std(medv, ddof=1) / np.sqrt(n)
print(f"SE(mean) analytic: {se_mean_analytic:.3f}")

SE(mean) analytic: 0.409


Interpretation:

This value quantifies the uncertainty in our estimate of the population mean 𝜇^. It tells us that if we repeatedly sampled 100 observations from the same population and computed the mean each time, the sample means would typically vary by about 0.409 units around the true population mean.

QUESTION C

Now estimate the standard error of µˆ using the bootstrap. How does this compare to your answer from (b)?

In [4]:
rng = np.random.default_rng(0)
B = 1000
boot_means = np.empty(B)

for b in range(B):
    sample = rng.choice(medv, size=n, replace=True)
    boot_means[b] = np.mean(sample)

se_mean_boot = np.std(boot_means, ddof=1)

print(f"SE(mean) bootstrap: {se_mean_boot:.3f}")


SE(mean) bootstrap: 0.413


The standard error estimate of µˆ using bootstrap give a slightly higher value than the non-bootstrapped one.

QUESTION D

Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of medv. Compare it to the results obtained by using Boston['medv'].std() and the two standard error rule (3.9).
Hint: You can approximate a 95 % confidence interval using the formula [ˆ µ − 2SE(ˆ µ), µˆ + 2SE(ˆ µ)].

In [5]:
ci_two_se = (mu_hat - 2*se_mean_boot, mu_hat + 2*se_mean_boot)
ci_percentile = (np.percentile(boot_means, 2.5), np.percentile(boot_means, 97.5))


print(f"95% CI (two-SE rule): [{ci_two_se[0]:.3f}, {ci_two_se[1]:.3f}]")
print(f"95% CI (bootstrap percentile): [{ci_percentile[0]:.3f}, {ci_percentile[1]:.3f}]")


95% CI (two-SE rule): [21.707, 23.358]
95% CI (bootstrap percentile): [21.767, 23.368]


They are almost overlapping and narrow, indicating a precise estimate of the population mean. 

QUESTION E

Based on this data set, provide an estimate, µˆmed, for the median value of medv in the population.

In [6]:
mu_med_hat = np.median(medv)

print(f"Median (mu_med_hat): {mu_med_hat:.3f}")

Median (mu_med_hat): 21.200


QUESTION F

We now would like to estimate the standard error of µˆmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [7]:
boot_medians = np.empty(B)
for b in range(B):
    sample = rng.choice(medv, size=n, replace=True)
    boot_medians[b] = np.median(sample)

se_median_boot = np.std(boot_medians, ddof=1)
ci_median_percentile = (np.percentile(boot_medians, 2.5), np.percentile(boot_medians, 97.5))

print(f"SE(median) bootstrap: {se_median_boot:.3f}")
print(f"95% CI median (bootstrap percentile): [{ci_median_percentile[0]:.3f}, {ci_median_percentile[1]:.3f}]")

SE(median) bootstrap: 0.378
95% CI median (bootstrap percentile): [20.500, 21.900]


Mean (22.533) and median (21.200) are close, indicating a normality and robustness to outliers. The confidence interval of median and mean overlap just a little.

QUESTION G

Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity µˆ0.1. (You can use the np.percentile() function.)

In [8]:
mu_p10_hat = np.percentile(medv, 10)

print(f"10th percentile (mu_p10_hat): {mu_p10_hat:.3f}")

10th percentile (mu_p10_hat): 12.750


QUESTION H

Use the bootstrap to estimate the standard error of µˆ0.1. Comment on your findings.

In [9]:
boot_p10 = np.empty(B)
for b in range(B):
    sample = rng.choice(medv, size=n, replace=True)
    boot_p10[b] = np.percentile(sample, 10)

se_p10_boot = np.std(boot_p10, ddof=1)
ci_p10_percentile = (np.percentile(boot_p10, 2.5), np.percentile(boot_p10, 97.5))

print(f"SE(10th percentile) bootstrap: {se_p10_boot:.3f}")
print(f"95% CI 10th percentile (bootstrap percentile): [{ci_p10_percentile[0]:.3f}, {ci_p10_percentile[1]:.3f}]")

SE(10th percentile) bootstrap: 0.498
95% CI 10th percentile (bootstrap percentile): [11.700, 13.450]


Interpretation:

SE higher than both mean and median, which means it has higher variability. It's expected, since its position is in the distribution's tail (spread wider for less amount of data).